File: python/025. Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        parent = s = ListNode(0)
        node = s.next = head        
        while node != None:
            cnt = 0
            while True:
                node = node.next
                cnt += 1
                if cnt == k or node == None: break
            
            if cnt == k:
                # do reverse
                t = node
                temp = s.next
                self.reverseList(s,t, k)
                s = temp
        return parent.next

    def reverseList(self, s, t, k):
        node = s.next
        for i in range(k):
            temp = node.next
            node.next = t
            t = node
            node = temp
        s.next = t

File: C++/025. Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head==NULL || k<=1){return head;}
        ListNode* dummy = new ListNode(0);
        dummy->next=head;
        ListNode* prev=dummy, *next, *node=head, *start, *end, *temp;
        while(node)
        {
            for (int i=0;i<k;i++)
            {
                if (node==NULL){return dummy->next;}
                end = node;
                node=node->next;
            }
            next=node;
            start = prev->next;
            node=start;
            for (int i=0;i<k-1;i++)
            {
                temp=node;
                prev->next = prev->next->next;
                node = prev->next;
                temp->next=end->next;
                end->next=temp;
            }
            node = next;
            prev = start;
        }
        return dummy->next;
    }
};
